### DISCUSSION TITLE
A Few questions on tables, filter, etc

### DISCUSSION THREAD

**Discussion by codemachine999** [Mimetis/Dotmim.Sync#792] [Q&A]
I have some questions about the tables:
1. Where do you place the list of tables to sync in web mode: client or server? (Your note says server).
2. If you need to filter rows, where do you place the filter (example suggests client)
3. If you place the tables on the server, where do you place the filters and parameters? Examples suggest client
4. How do you get the values of the filters to the server if you place them on the client?

**Comment by Mimetis** [comment]
> 1. Where do you place the list of tables to sync in web mode: client or server? (Your note says server).

The tables definition is contained in the `SyncSetup` instance.
This definition is defined on the server side.
Definition can contain: Table names, columns names, and filters

This definition is downloaded by the client each time a sync is starting.
The definition downloaded is saved locally, and is compared to the server definition on each start
It allows `DMS` to ensure the `SyncSetup` did not change between sync (An exception is raised is SyncSetup has changed between 2 syncs)

For example:
``` cs

services.AddControllers();
services.AddDistributedMemoryCache();
services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));

 // Create the setup used for your sync process
 var tables = new string[] {"ProductCategory",
     "ProductModel", "Product",
     "Address", "Customer", "CustomerAddress",
     "SalesOrderHeader", "SalesOrderDetail" };

 var setup = new SyncSetup(tables)
 {
     // optional :
     StoredProceduresPrefix = "s",
     StoredProceduresSuffix = "",
     TrackingTablesPrefix = "s",
     TrackingTablesSuffix = ""
 };

services.AddSyncServer<SqlSyncProvider>(connectionString, setup, options);
```

> 2. If you need to filter rows, where do you place the filter (example suggests client)

The **definition** of the filter and the parameters are placed on the server side, in the `SyncSetup` instance:

Definition is:
- Parameter name
- Parameter type
- Filter clause

``` cs
// Here we have a filter on Address with 2 parameters
var addressFilter = new SetupFilter("Address");
addressFilter.AddParameter("City", "Address", true);
addressFilter.AddParameter("postal", DbType.String, true, null, 20);
addressFilter.AddWhere("City", "Address", "City");
addressFilter.AddWhere("PostalCode", "Address", "postal");
setup.Filters.Add(addressFilter);

services.AddSyncServer<SqlSyncProvider>(connectionString, setup, options);
```

The **values** of the parameters are coming from each client, on the client side:

``` cs

var serverOrchestrator = new WebRemoteOrchestrator("https://localhost:44342/api/sync");
var clientProvider = new SqlSyncProvider(clientConnectionString);
var agent = new SyncAgent(clientProvider, serverOrchestrator, options);

 var parameters = new SyncParameters
 {
     { "City", "Toronto" },
     // Because I've specified that "postal" could be null, 
     // I can set the value to DBNull.Value (and the get all postal code in Toronto city)
     { "postal", DBNull.Value }
 };

// Launch the sync process
var s1 = await agent.SynchronizeAsync(progress);
// Write results
Console.WriteLine(s1);

```

> 3. If you place the tables on the server, where do you place the filters and parameters? Examples suggest client

See previous answers

> 4. How do you get the values of the filters to the server if you place them on the client?

They are embedded in the JSON payload on each call
